experimental/rel_component/scripts/rel_model.py

```diff
@@ -0,0 +1,133 @@
+from typing import List, Callable, Tuple, Optional
+from thinc.types import Floats2d
+from thinc.api import Model, Linear, Ops, chain, Logistic
+
+from spacy.util import registry
+
+
+@registry.architectures.register("rel_model.v1")
+def create_relation_model(
+    tok2vec: Model[List["Doc"], List[Floats2d]],
+    get_candidates: Callable[["Doc"], List[Tuple["Span", "Span"]]],
+    create_candidate_tensor: Callable[[List["Doc"], Floats2d, Ops], Tuple[Floats2d, Callable]],
+    output_layer: Model[Floats2d, Floats2d],
+    nO: Optional[int],
+) -> Model[List["Doc"], Floats2d]:
+    return Model(
+        "relations",
+        layers=[tok2vec, output_layer],
+        refs={"tok2vec": tok2vec, "output_layer": output_layer},
+        attrs={
+            "create_candidate_tensor": create_candidate_tensor,
+            "get_candidates": get_candidates,
+        },
+        dims={"nO": nO},
+        forward=forward,
+        init=init,
+    )
+
+
+@registry.misc.register("rel_cand_tensor.v1")
+def create_tensors() -> Callable[[List["Doc"], Callable, List[Floats2d], Ops], Tuple[Floats2d, Callable]]:
+    def get_candidate_tensor(docs: List["Doc"], get_candidates: Callable, tokvecs: List[Floats2d], ops: Ops):
```

Comment:
I think the best way to do this will be to build a `Ragged` array of the vectors for the entity tokens, and then you'll be able to use the `reduce_mean` layer. It'll be quite similar to the code in `spacy-transformers`.

Let's say you have three entities, at spans `(0, 3), (10, 11), (12, 18)`, and your token vectors are of width 128. You'd have a `Ragged` array where the data is of shape `(7, 128)` and the lengths are a `Ints1d` with contents `[3, 1, 6]`. You'd also store the indices array [0, 1, 2, 10, 12, 13, 14, 15, 16, 17]` so that you can calculate the gradients.

Reply:
That makes sense to me. I guess I worry a little about readability of the code though? The `Ragged` datastructure does take some getting used to. The same's true for using `scatter_add` for instance, which I think obscures things a little? If we want to use this example as a sort of tutorial?

Reply:
Not sure if my comment is relevant (if so then ignore me)

From a user perspective it would be great to see both. So we have the readable solution and the more production ready solution. Which solution to actually use will be determined by the config I presume.

Reply:
I think it probably won't be less readable, because you'll be doing a bit less in each step. Having an explicit representation of the entities will be good.

If you don't want to build the `Ragged` array for the example, I would recommend building a `List[Floats2d]` for the entities instead, and then chaining a `list2ragged` layer before the pooling.

Reply:
Something like:

```python

def instance_forward(
    model: Model[List[Doc], Floats2d], docs: List[Doc], is_train: bool
) -> Tuple[Floats2d, Callable]:
    pooling: Model[List[Floats2d], Floats2d] = model.get_ref("pooling")
    tok2vec: Model[List[Doc], List[Floats2d]] = model.get_ref("tok2vec")
    tokvecs, bp_tokvecs = tok2vec(docs, is_train)
    get_instances = model.attrs["get_instances"]
    instances = [get_instances(doc) for doc in docs]
    for (ent1, ent2), tokvec in zip(instances, tokvecs):
        ents.append(tokvec[ent1.start : ent1.end])
        ents.append(tokvec[ent2.start : ent2.end])
    pooled, bp_pooled = pooling(ents, is_train)
    # Reshape so that pairs of rows are concatenated.
    # We need the entries at rows (0, 1), (2, 3) etc to be one vector.
    relations = model.ops.reshape2f(pooled, -1, pooled.shape[1] * 2)

    def backprop(d_relations: Floats2d) -> List[Doc]:
        d_pooled = model.ops.reshape2f(d_relations, d_relations.shape[0] * 2, -1)
        d_ents = bp_pooled(d_pooled)
        d_tokvecs = []
        for ents1, ents2 in instances:
            i = len(d_tokvecs)
            d_tokvec = model.ops.alloc2f(len(docs[i]), d_pooled.shape[1])
            d_tokvec[ents1.start : ents1.end] += d_ents[i]
            d_tokvec[ents2.start : ents2.end] += d_ents[i+1]
            d_tokvecs.append(d_tokvec)
        d_docs = bp_tokvecs(d_tokvecs)
        return d_docs

    return relations, backprop
```

Reply:
Ok I ended up trying both ways, and I agree with Matt's original idea of using `Ragged`. In the video/tutorial that we'll record for this, we can explain the `Ragged` data type as it's also an important datastructure to understand the alignment in `spacy-transformers`. The code is much nicer now :-)